### PR TITLE
[feature-wip](nereids) Make nereids more compatible with spark-sql syntax.

### PR DESCRIPTION
**Pull request by dutyu**
## Proposed changes

**Thanks for** pr #21855 to provide a wonderful reference. 

Maybe it is very difficult and **cost-expensive** to implement **a comprehensive logical plan adapter**, maybe there is just some small syntax variations between doris and some other engines (such as hive/spark), so we can just **focus on** the **difference** here.

This pr mainly focus on the **syntax difference between doris and spark-sql**. For instance, do some function tranformations and override some syntax validations.

- add a dialect named `spark_sql`
- move method `NereidsParser#parseSQLWithDialect` to `TrinoParser`
- extract some `FnCallTransformer`/`FnCallTransformers` classes, so we can reuse the logic about the function transformers
- allow derived tables without alias when we set dialect to `spark_sql`(legacy and nereids parser are both supported)
- add some function transformers for hive/spark built-in functions

### Test case (from our online doris cluster)

- Test derived table without alias

```sql
MySQL [(none)]> show variables like '%dialect%';
+---------------+-------+---------------+---------+
| Variable_name | Value | Default_Value | Changed |
+---------------+-------+---------------+---------+
| sql_dialect   | spark_sql  | doris         | 1       |
+---------------+-------+---------------+---------+
1 row in set (0.01 sec)

MySQL [(none)]> select * from (select 1);
+------+
| 1    |
+------+
|    1 |
+------+
1 row in set (0.03 sec)

MySQL [(none)]> select __auto_generated_subquery_name.a from (select 1 as a);
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.03 sec)

MySQL [(none)]> set sql_dialect=doris;
Query OK, 0 rows affected (0.02 sec)

MySQL [(none)]> select * from (select 1);
ERROR 1248 (42000): errCode = 2, detailMessage = Every derived table must have its own alias
MySQL [(none)]> 
```

- Test spark-sql/hive built-in functions

```sql
MySQL [(none)]> show global functions;
Empty set (0.01 sec)

MySQL [(none)]> show variables like '%dialect%';
+---------------+-------+---------------+---------+
| Variable_name | Value | Default_Value | Changed |
+---------------+-------+---------------+---------+
| sql_dialect   | spark_sql  | doris         | 1       |
+---------------+-------+---------------+---------+
1 row in set (0.01 sec)

MySQL [(none)]> select get_json_object('{"a":"b"}', '$.a');
+----------------------------------+
| json_extract('{"a":"b"}', '$.a') |
+----------------------------------+
| "b"                              |
+----------------------------------+
1 row in set (0.04 sec)

MySQL [(none)]> select split("a b c", " ");
+-------------------------------+
| split_by_string('a b c', ' ') |
+-------------------------------+
| ["a", "b", "c"]               |
+-------------------------------+
1 row in set (1.17 sec)
```

### CODE REVIEW

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/InlineViewRef.java [hunk 1]
```diff
@@ -186,7 +189,13 @@ public void analyze(Analyzer analyzer) throws AnalysisException, UserException {
         }
 
         if (view == null && !hasExplicitAlias()) {
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/InlineViewRef.java:191]: this is only for old optimizer, if support new optimizer, should also add logical in `org.apache.doris.nereids.parser.LogicalPlanBuilder#visitAliasedQuery`

Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/InlineViewRef.java:191]: if meet inline view without alias, return '_', is this consistent to hive ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/NereidsParser.java [hunk 2]
```diff
@@ -59,45 +58,45 @@ public class NereidsParser {
...
      * ParseSQL with dialect.
      */
     public List<StatementBase> parseSQL(String sql, SessionVariable sessionVariable) {
-        if (ParseDialect.TRINO_395.getDialect().getDialectName()
-                .equalsIgnoreCase(sessionVariable.getSqlDialect())) {
-            return parseSQLWithDialect(sql, sessionVariable);
-        } else {
-            return parseSQL(sql);
+        ParseDialect.Dialect sqlDialect = ParseDialect.Dialect.getByName(sessionVariable.getSqlDialect());
+        if (sqlDialect != null) {
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/NereidsParser.java:69]: sqlDialect seems never is null, can not reach the below code `parseSQL(sql)`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/ParseDialect.java [hunk 3]
```diff
@@ -58,7 +62,9 @@ public enum Version {
         /**
          * Doris parser and it's version is 2.0.0.
          */
-        DORIS_2_ALL("2.*");
+        DORIS_2_ALL("2.*"),
+
+        HIVE_ALL("*");
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/ParseDialect.java:67]: Should add java doc

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/ParseDialect.java [hunk 4]
```diff
@@ -81,7 +87,9 @@ public enum Dialect {
         /**
          * Doris parser dialect
          */
-        DORIS("doris");
+        DORIS("doris"),
+
+        HIVE("hive");
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/ParseDialect.java:92]: Should add java doc

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/ParseDialect.java [hunk 5]
```diff
@@ -29,7 +29,11 @@ public enum ParseDialect {
     /**
      * Doris parser and it's version is 2.0.0.
      */
-    DORIS_2_ALL(Dialect.DORIS, Version.DORIS_2_ALL);
+    DORIS_2_ALL(Dialect.DORIS, Version.DORIS_2_ALL),
+    /**
+     * Hive parser.
+     */
+    HIVE_ALL(Dialect.HIVE, Version.HIVE_ALL);
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/ParseDialect.java:36]: Use HIVE_ALL, the high and low version hive dialect is compatible? should split them to hive2 , hive3?

**Comment by dutyu** [comment]
Use dialect `spark_sql` instead of `hive`, in fact `Spark SQL` support a subquery without a explict alias, but hive does not support.

Reference: 
[SubQuery syntax of Hive](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+SubQueries)

The docs of `Spark SQL` does not describe the specific syntax about subquery, but we can find it at source code.

[spark-sql#visitAliasedQuery](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L1689)  

[spark-sql#generateSubqueryName](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L1746)

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by seawinde — APPROVED** [review]

**Pull request merged**